src/core/tests/logger.cpp

```diff
@@ -0,0 +1,162 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+# include <gtest/gtest.h>
+# include <dsn/service_api_cpp.h>
+# include <dsn/service_api_c.h>
+# include <dsn/tool/providers.hpc.h>
+# include <dsn/internal/singleton_store.h>
+# include <dsn/internal/factory_store.h>
+# include <dsn/tool_api.h>
+# include <dsn/internal/logging_provider.h>
+
+# include "service_engine.h"
+# include "test_utils.h"
+# include "../tools/hpc/hpc_logger.h"
+# include "../tools/hpc/hpc_tail_logger.h"
+# include "../tools/common/simple_logger.h"
+
+
+
+
+using namespace ::dsn;
+const char str[64] = "this is a logging test for log %010d @ thread %010d";
+
+template<typename TLOGGER>
+void logv(TLOGGER &logger,const char* fmt, ...)
+{
+	va_list ap;
+	va_start(ap, fmt);
+	logger.dsn_logv(
+		__FILE__,
+		__FUNCTION__,
+		__LINE__,
+		LOG_LEVEL_DEBUG,
+		"log.test",
+		str,
+		ap
+		);
+	va_end(ap);
+}
+
+template<typename TLOGGER>
+void logger_test(int thread_count, int record_count)
+{
+	std::list<std::thread*> threads;
+	TLOGGER logger;
+
+	uint64_t nts = dsn_now_ns();
+	uint64_t nts_start = nts;
+
+	
+	for (int i = 0; i < thread_count; ++i)
+	{
+		threads.push_back(new std::thread([&,i]
+		{
+			task::set_tls_dsn_context(task::get_current_node2(), nullptr, nullptr);
+			for (int j = 0; j < record_count; j++)
+			{
+				logv(logger, str, j, i);
+			}
+			
+		}));
+	}
+	
+	for (auto& thr : threads)
+	{
+		thr->join();
+		delete thr;
+	}
+	threads.clear(); 
+
+	nts = dsn_now_ns();
+	
+	//one sample log
+	size_t size_per_log = strlen(
+		"13:11:02.678 (1446037862678885017 1d50) unknown.io-thrd.07504: this is a logging test for log 0000000000 @ thread 000000000"
+		)+1;
+
+	std::cout 
+		<< thread_count << "\t\t\t "
+		<< record_count << "\t\t\t " 
+		<< static_cast<double>(thread_count * record_count) 
+			* size_per_log / (1024*1024) / (nts - nts_start) * 1000000000
+		<< "MB/s" 
+		<< std::endl;
+
+	//logger.flush();
+
+}
+
+TEST(core, hpc_logger_test)
+{
+	std::cout << "thread_count\t\t record_count\t\t speed" << std::endl;
+
+	auto threads_count = {1,2,4,10,20};
+	auto record_count = { 10000,100000,200000,500000,1000000};
```

Comment:
there is no need for various record counts as it does not reveal more information than one case. u may simply pick one that runs enough time, e.g., > 20 s.